clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -529,13 +603,21 @@ private ListOffsetResult fetchOffsetsByTimes(Map<TopicPartition, Long> timestamp
     }
 
     private List<ConsumerRecord<K, V>> fetchRecords(PartitionRecords partitionRecords, int maxRecords) {
+        boolean shouldDrainRecords = true;
         if (!subscriptions.isAssigned(partitionRecords.partition)) {
             // this can happen when a rebalance happened before fetched records are returned to the consumer's poll call
             log.debug("Not returning fetched records for partition {} since it is no longer assigned",
                     partitionRecords.partition);
         } else if (!subscriptions.isFetchable(partitionRecords.partition)) {
             // this can happen when a partition is paused before fetched records are returned to the consumer's
             // poll call or if the offset is being reset
+            // check if the TopicPartition is assigned but is paused. If it is paused, buffer the fetched records for
+            // reuse on unpause
+            if (subscriptions.isPaused(partitionRecords.partition)) {
+                this.pausedNextInLineRecordsPerTopicPartition.put(partitionRecords.partition, partitionRecords);
+                shouldDrainRecords = false;
+                nextInLineRecords = null;
+            }
             log.debug("Not returning fetched records for assigned partition {} since it is no longer fetchable",
```

Comment:
Maybe put a log statement, indicating that we are putting some data aside in the `pausedNextInLineRecordsPerTopicPartition` plus its size.